File: Longest Substring without Repeating Characters/Leetcode_3.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.length() == 0) return 0;
        int maxLen=1;
        string cur;
        for (char& c: s)
            if (cur.find_first_of(c) == string::npos){
                cur += c;
                if (cur.length() > maxLen) maxLen=cur.length();
            } else {
                size_t pos = cur.find_first_of(c);
                cur = cur.substr(++pos) + c;
            }
        return maxLen;
    }
};

File: Longest Substring without Repeating Characters/Leetcode_3.py
# store current valid substring as str
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s: return 0
        cur, maxLen = "", 0
        for char in s:
            if char not in cur:
                cur += char
                maxLen = max(maxLen, len(cur))
            else:
                removed = cur.split(char)
                cur = removed[1] + char
        return maxLen
    
 # store current valid substring in a dict
 class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = 1 if s else 0
        cur, start = {}, 0
        for i, char in enumerate(s):
            if char in cur:
                j = cur.pop(char)
                cur = {char2: k  for char2, k in cur.items() if k>j}
                start = min(cur.values()) if cur else i
            cur[char] = i
            res = max(res, i - start + 1)
        return res
